### PR TITLE
fix: preserve search params in "onUnhandledRequest" messages

### LINKED ISSUE #2125: Preserve the original request URL in "onUnhandledRequest" warnings
### Prerequisites

- [X] I confirm my issue is not in the [opened issues](https://github.com/mswjs/msw/issues)
- [X] I confirm the [Frequently Asked Questions](https://mswjs.io/docs/faq) didn't contain the answer to my issue

### Environment check

- [X] I'm using the [latest](https://github.com/mswjs/msw/releases/latest) `msw` version
- [X] I'm using Node.js version 18 or higher

### Node.js version

v20.11.0

### Reproduction repository

Always specify repros. Don't follow my lead.

### Reproduction steps

1. Set up MSW in the browser.
2. Configure the `onUnhandledRequest` callback:
```ts
await worker.start({
  onUnhandledRequest(request, print) {
    console.log('URL:', request.url)
    if (/\.(css|js|json|png|jpg|gif)$/.test(request.url)) {
      return
    }

    print.warning()
  },
})
```
3. Make a request like `/fo/bar/?t=123`

### Current behavior

MSW prints a clean URL, making it hard to understand why my custom `onUnhandledRequest` logic is not matching this request:

Screenshot 2024-04-05 at 15 45 08

### Expected behavior

MSW preserves the unhandled request URL as-is, which includes keeping its query parameters and fragments.

### PR DESCRIPTION
**Pull request by nagadevkrishna**
- Fixes https://github.com/mswjs/msw/issues/2125

### CODE REVIEW

File: src/core/utils/request/toPublicUrl.ts [hunk 1]
```diff
@@ -9,7 +9,9 @@ export function toPublicUrl(url: string | URL): string {
 
   const urlInstance = url instanceof URL ? url : new URL(url)
 
+  const [, relativeUrl] = urlInstance.href.split(urlInstance.origin)
```
Comment (kettanaito) [inline src/core/utils/request/toPublicUrl.ts]: This does solve the problem but it changes the other use cases we have for `toPublicUrl`. 

I think a far better approach is to keep these changes scoped to the immediate problem area: the `toPublic()` usage in the `onUnhandledRequest.ts`. 

https://github.com/mswjs/msw/blob/e83a6699bc40a4fc2a593a93346158b8b8b61649/src/core/utils/request/onUnhandledRequest.ts#L24-L27

Let's try something like this:

```ts
const url = new URL(request.url)
const publicUrl = toPublicUrl(url) + url.search
```

This way both relative and absolute URLs will be printed correctly _but_ will also have any search parameters preserved. 

Then, we can modify the tests in `onUnhandledRequest.test.ts` to include two new scenarios:

- Unhandled request with a relative URL and search params;
- Unhandled request with an absolute URL and search params.

**Review by kettanaito — CHANGES_REQUESTED** [review]
Thanks for tackling this, @nagadevkrishna! I left a suggestion, let me know what you think about it.

File: src/core/utils/request/toPublicUrl.ts (hunk 1 above)
Comment (nagadevkrishna) [reply src/core/utils/request/toPublicUrl.ts]: Thanks for the suggestion. Made changes now and pushed to PR again.

**Review by kettanaito — APPROVED** [review]
This looks fantastic!

**Comment by nagadevkrishna** [comment]
awesome!

**Comment by kettanaito** [comment]
Don't close it just yet. I need to merge it so it gets to MSW. Be patient.

**Comment by kettanaito** [comment]
Welcome to the contributors, @nagadevkrishna! 👏

**Comment by kettanaito** [comment]
## Released: v2.2.14 🎉

This has been released in v2.2.14!

- 📄 [**Release notes**](https://github.com/mswjs/msw/releases/tag/v2.2.14)
- 📦 [npm package](https://www.npmjs.com/package/msw/v/2.2.14)

Make sure to always update to the latest version (`npm i msw@latest`) to get the newest features and bug fixes.

---

_Predictable release automation by [@ossjs/release](https://github.com/ossjs/release)_.

**Pull request merged**